Improved column Resize

jet11x opened the pull request:
## Related Issue

#400 

## Description

Support two forms of resizable columns.
1. Full width table - similar to before but with only two columns changing size not lots of them.
2. Variable width table - alter one columns width at a time

added props:
- onColumnResize callback
- tableWidth 'full' (default) and 'variable'

## Related PRs

## Impacted Areas in Application

List general components of the application that this PR will affect:

Headers, table width and column width.

\*

## Additional Notes

Changes, thoughts and questions.

MTableHeader/index.js
  - need to change state and whilst doing so changed to more normal function component pattern
  - contrained widths to min/max.  Have default of 20 and 10000 for min/max - could expose this as props
  - change sort button to not surround title.  This was to allow ellipsis for title
  - changed so works with no Resize icon, useful if using cell borders (alhough perhaps not needed after I sorted alignment)
  - resize zone is equal amount to left and right of icon/border
  - some more JSS/CSS classes
  
data-manager.js
  - added tableStyleWidth to record width in css style

function setTableWidth
  - commented out following line as stopped usedWidth from bring calculated correctly.
  - `columnDef.tableData.width !== undefined`
    
funnction onColumnResized changed to deal with two ways of resizing (changing one column or two)
  - should most of this be elsewhere so data-manager is just state/context?  Similar for setTableWidth?
  
material-table.js
  - support onColumnResized prop
  - in reviewing code I spotted I've added fn getInitialColumnWidth which I need to remove

index.d.ts - would explicit Type for column info from onColumnResized be a good idea?

demo-components\index.js
  - Added new example ResizableTableWidthVariable
  - May be worth having demo specific to column resize

jet11x committed:
Improved column Resize

matthewoestreich reviewed (approved):
Nice work. Everything looks straight forward and well done. Love the comments, too! Thank you very much for your contribution.

Before merging, I would like @Domino987 to review as well, just because it's a nice sized PR.

What do you think @Domino987 ?

jet11x commented:
Thanks for the very fast response.  As you say a fair sized PR, so more
than happy to wait for a further review.

On Wed, 22 Dec 2021 at 18:16, Matt Oestreich ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> Nice work. Everything looks straight forward and well done. Love the
> comments, too! Thank you very much for your contribution.
>
> Before merging, I would like @Domino987 <https://github.com/Domino987> to
> review as well, just because it's a nice sized PR.
>
> What do you think @Domino987 <https://github.com/Domino987> ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/material-table-core/core/pull/444#pullrequestreview-838725511>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO6M2T5X3N2SE7PZ7CWYPDUSIIXZANCNFSM5KS6TMVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jet11x commented:
There might be a problem, I'm seeing an error using material-table from
another React app.
Error:  invalid hook call. Hooks can only be called inside of the body of a
function component
I'm investigating.


On Wed, 22 Dec 2021 at 18:21, John Talintyre ***@***.***> wrote:

> Thanks for the very fast response.  As you say a fair sized PR, so more
> than happy to wait for a further review.
>
> On Wed, 22 Dec 2021 at 18:16, Matt Oestreich ***@***.***>
> wrote:
>
>> ***@***.**** approved this pull request.
>>
>> Nice work. Everything looks straight forward and well done. Love the
>> comments, too! Thank you very much for your contribution.
>>
>> Before merging, I would like @Domino987 <https://github.com/Domino987>
>> to review as well, just because it's a nice sized PR.
>>
>> What do you think @Domino987 <https://github.com/Domino987> ?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/material-table-core/core/pull/444#pullrequestreview-838725511>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACO6M2T5X3N2SE7PZ7CWYPDUSIIXZANCNFSM5KS6TMVA>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

matthewoestreich commented:
@jet11x were you able to resolve this? We may need to write tests for this.

jet11x commented:
I haven't resolved, but I get the same problem using head and referring to
it locally.  So possible problem is how I'm locally referencing and not the
changes I made.

On Wed, 22 Dec 2021 at 22:20, Matt Oestreich ***@***.***>
wrote:

> @jet11x <https://github.com/jet11x> were you able to resolve this? We may
> need to write tests for this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/material-table-core/core/pull/444#issuecomment-999915035>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO6M2RQJXVXBZQ7CVXMC2LUSJFKBANCNFSM5KS6TMVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jet11x commented:
When I used yarn link it diesn't work.
When I created tarball with yarn pack it worked fine.
So all good now.


On Thu, 23 Dec 2021 at 06:32, John Talintyre ***@***.***> wrote:

> I haven't resolved, but I get the same problem using head and referring to
> it locally.  So possible problem is how I'm locally referencing and not the
> changes I made.
>
> On Wed, 22 Dec 2021 at 22:20, Matt Oestreich ***@***.***>
> wrote:
>
>> @jet11x <https://github.com/jet11x> were you able to resolve this? We
>> may need to write tests for this.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/material-table-core/core/pull/444#issuecomment-999915035>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACO6M2RQJXVXBZQ7CVXMC2LUSJFKBANCNFSM5KS6TMVA>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

jet11x commented:
I'm making some improvements (and let's be honest I've fixed a couple of
bugs).
So will hold off on merge for now.

On Thu, 23 Dec 2021 at 06:49, John Talintyre ***@***.***> wrote:

> When I used yarn link it diesn't work.
> When I created tarball with yarn pack it worked fine.
> So all good now.
>
>
> On Thu, 23 Dec 2021 at 06:32, John Talintyre ***@***.***> wrote:
>
>> I haven't resolved, but I get the same problem using head and
>> referring to it locally.  So possible problem is how I'm locally
>> referencing and not the changes I made.
>>
>> On Wed, 22 Dec 2021 at 22:20, Matt Oestreich ***@***.***>
>> wrote:
>>
>>> @jet11x <https://github.com/jet11x> were you able to resolve this? We
>>> may need to write tests for this.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/material-table-core/core/pull/444#issuecomment-999915035>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/ACO6M2RQJXVXBZQ7CVXMC2LUSJFKBANCNFSM5KS6TMVA>
>>> .
>>> Triage notifications on the go with GitHub Mobile for iOS
>>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>>> or Android
>>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>>
>>> You are receiving this because you were mentioned.Message ID:
>>> ***@***.***>
>>>
>>

Domino987 commented on src/components/MTableHeader/index.js:
```diff
@@ -168,60 +270,22 @@ export function MTableHeader({ onColumnResized, ...props }) {
                   }
                 >
                   {columnDef.title}
+                  {renderSortButton(columnDef)}
```
I am thinki if this could be its own react component to avoid remounts?

jet11x replied:
That's done - I've left in the same file

Domino987 commented on src/components/MTableHeader/index.js:
```diff
@@ -10,52 +10,101 @@ import { Tooltip, withStyles } from '@material-ui/core';
+    // Extra max/min are to avoid sudden column changes when a column that starts without
+    // an explicit width is resized
+    const constrainedNewWidth = Math.min(
+      Math.max(col.maxWidth || 10000, lastWidth), // Avoid sudden decrease in column width
```
Could we move those to constants outside of the component to reduce the amount of magic numbers and make it easier to extract them to props later?

jet11x replied:
I wasn't happy with that either!
Moved to constants - I left in it, as that felt step closer to adding as prop later.

Domino987 reviewed (changes requested):
This is awesome and looks great. Thank you for this. I just have 2 remarks

jet11x committed:
Fixed a few issues and tidied up
Added tests for column Resize

jet11x commented:
Whilst adding tests I noticed that data-manager alters props by adding tableData into each column.  Pretty sure this isn't good practise for React components.  See plans include moving from data-manager to ContextProvider - good opportunity then to 'fix' this?

Domino987 commented:
We are making copies of both the rows and the columns to add the tabledata as far as I know. Or where do you mean?

jet11x commented:
This is shown in one of the tests I added.  Just checked I hadn't
introduced the problem by changing an existing test on master.

test('renders summary row if renderSummaryRow prop is present', () => {
  render(
    <MaterialTable
      data={data}
      columns={columns}
      renderSummaryRow={({ column }) => `Summary_${column.title}`}
    />
  );

  columns.forEach(col => {
    console.log('col: ', col)
  })

  expect(
    screen.getByRole('row', { name: 'Summary_Enum Summary_Name' })
  ).toBeTruthy();
});

This gives the output:

col:  {
      title: 'Name',
      field: 'id',
      tableData: {
        columnOrder: 1,
        filterValue: undefined,
        groupOrder: undefined,
        groupSort: 'asc',
        width: 'calc((100% - (0px)) / 2)',
        initialWidth: 'calc((100% - (0px)) / 2)',
        additionalWidth: 0,
        id: 1
      }
    }


On Sun, 2 Jan 2022 at 00:09, Domino987 ***@***.***> wrote:

> We are making copies of both the rows and the columns to add the tabledata
> as far as I know. Or where do you mean?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/material-table-core/core/pull/444#issuecomment-1003639540>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO6M2WI7E37OYCIEJ326VTUT6JVPANCNFSM5KS6TMVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Domino987 commented:
Yes it returns the new column/data bit this is not the same object. It's a copy.

jet11x commented:
This may be beyond my understanding of React/JS, but given prop reference
is passed in we can't have a copy returned, the array of columns is being
altered rather than a copy taken of the array.  Feels like an anti-pattern
with internals of Material Table exposed to the user of the component.  I'm
assuming it works a React does do a deep equality check.

On Mon, 3 Jan 2022 at 09:23, Domino987 ***@***.***> wrote:

> Yes it returns the new column/data bit this is not the same object. It's a
> copy.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/material-table-core/core/pull/444#issuecomment-1003963055>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO6M2Q4M2EOK2WFRHCGWKTUUFTKVANCNFSM5KS6TMVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jet11x commented:
Anything more I need to look into or are you good with the changes I've made?

jet11x commented:
Thanks for reviewing and approving.
Been some time since I contributed to Open Source - great to be doing it again.

Domino987 commented:
@jet11x thank you for this 🙂 if you want to contribute more 😜 always good to have good devs working on the project!

jet11x commented:
Thanks @Domino987 - I could well be interested in some more contributions.
I think first should be the documentation for the change I've done.
A couple of other thoughts:
1. Do you want any help merging the changes I made to the next branch?
2. I'd also like to add drag/drop row re-ordering - first would be to play
with it a bit and get an idea of how extensive the change would be

On Mon, 10 Jan 2022 at 12:01, Domino987 ***@***.***> wrote:

> @jet11x <https://github.com/jet11x> thank you for this 🙂 if you want to
> contribute more 😜 always good to have good devs working on the project!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/material-table-core/core/pull/444#issuecomment-1008806693>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO6M2X2H7GY54TLUXXZKBTUVLC7ZANCNFSM5KS6TMVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Domino987 commented:
I already released a next version so this is done. Bit row reordering would be interesting. Go ahead!!

sachinagarwal010 commented:
@Domino987 @jet11x  I can see because of the fix provided as part of this until we don't set draggable as true Sorting will be hidden/disabled. Can you please check on this.

pratik-nagpure commented:
> @Domino987 @jet11x I can see because of the fix provided as part of this until we don't set draggable as true Sorting will be hidden/disabled. Can you please check on this.

True. It is forcing to set draggable as true to get sorting on column.

jet11x commented:
Taken initial look.  I get it, if option, draggable is false, sorting icon not there in header.  I can see RenderSortButton is inside the Draggable code which include columnDef.draggable !== false.  Should be easy to pull it out - not sure how I managed to change that ... 
I've got fix for this.  Do you want a new pull request?  An issue created?

On Mon, 24 Jan 2022 at 08:41, Pratik Nagpure ***@***.***>
wrote:

> @Domino987 <https://github.com/Domino987> @jet11x
> <https://github.com/jet11x> I can see because of the fix provided as part
> of this until we don't set draggable as true Sorting will be
> hidden/disabled. Can you please check on this.
>
> True. It is forcing to set draggable as true to get sorting on column.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/material-table-core/core/pull/444#issuecomment-1019848517>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO6M2XAPZJK74VW5KPBXNLUXUGFJANCNFSM5KS6TMVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
